#Execrise 4.8 walk through

In [2]:
#Importing libraries 
import pandas as pd
import numpy as np
import os

#path setup

In [3]:
path=r"C:\Users\ranji\Downloads\Submitted tasks\Python_5\Instacart\Instacart-Basket-Analysis\02 Data\Prepared Data"

In [4]:
ords_prods_merge=pd.read_pickle(os.path.join(path,"orders_products_merged_updated.pkl"))

In [ ]:
ords_prods_merge.head()

In [30]:
df = ords_prods_merge[:1000000]

In [4]:
ords_prods_merge.shape

(32434212, 18)

In [6]:
ords_prods_merge.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   
5          1         Soda        77              7     9.0   both   
6          1         Soda        77              7     9.0   both   
7          1         Soda        77              7     9.0   both   
8          1         Soda        77              7     9.0   both   
9          1         Soda        77              7     9.0   both   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders  
5  Regularly busy   Regularly busy        Average orders  
6  Regularly busy     Busiest days           Most orders  
7  Regularly busy     Busiest days           Most orders  
8  Regularly busy     Busiest days           Most orders  
9      Least busy  Least busy days        Average orders


#Grouping Data with Pandas


In [7]:
ords_prods_merge.groupby('product_name')


#Aggregating Data with agg()


In [8]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#Different way of finding mean of 'order_number'

In [10]:
ords_prods_merge.groupby('department_id').order_number.mean()

department_id
1     15.457687
2     17.277920
3     17.179756
4     17.811403
5     15.213779
6     16.439806
7     17.225773
8     15.340520
9     15.895474
10    20.197148
11    16.170828
12    15.887622
13    16.583304
14    16.757377
15    16.165037
16    17.663250
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

In [16]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [18]:
ords_prods_merge.head(20)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   
15   2293453       19    prior             2                   5   
16   1973799       19    prior             5                   6   
17    532817       19    prior             7                   4   
18   1573906       21    prior            10                   3   
19   1593000       31    prior            10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
5           1         Soda        77              7     9.0   both   
6           1         Soda        77              7     9.0   both   
7           1         Soda        77              7     9.0   both   
8           1         Soda        77              7     9.0   both   
9           1         Soda        77              7     9.0   both   
10          0         Soda        77              7     9.0   both   
11          1         Soda        77              7   


#Deriving Columns with loc()


In [17]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [19]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [20]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [22]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

In [24]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [34]:
# check the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column

ords_prods_merge[['user_id', 'max_order', 'loyalty_flag']].sample(30)

user_id  max_order      loyalty_flag
17940016   188420         22  Regular customer
27979504     7806         28  Regular customer
3170380     74937          4      New customer
13544481   203748         39  Regular customer
132766      37699         32  Regular customer
19404161    15741         11  Regular customer
6253188     66918         69    Loyal customer
9571163     59272         29  Regular customer
8032776     12133         11  Regular customer
13432183    77281         12  Regular customer
32044240   196614          9      New customer
19898734    53383          3      New customer
10238566    90653          9      New customer
25559232   130276         10      New customer
25704959   149099         25  Regular customer
20789378   103028         59    Loyal customer
27323804   160093         48    Loyal customer
6046288    188020         99    Loyal customer
7985929     64402         17  Regular customer
7339671    132365         19  Regular customer
22624924   164046         13  Regular customer
26591285    77293         96    Loyal customer
24099589    69448         30  Regular customer
18408310    16475         38  Regular customer
7541050    171935         29  Regular customer
29604843   147181         25  Regular customer
1622610     41043         76    Loyal customer
9606270    170741         87    Loyal customer
10926482    43976         25  Regular customer
28033428    69932         12  Regular customer

Step 1 
Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

#Done


Step 2
In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [26]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379


Step 3
Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [32]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

The average order_number from the subset (1000000) is higher than the average order_number from the entire dataframe, excluding department_id: 17.


Step 4
Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [36]:
ords_prods_merge[['user_id', 'max_order', 'loyalty_flag']].sample(30)

user_id  max_order      loyalty_flag
24592334    48295         88    Loyal customer
2465645        27         81    Loyal customer
31225002   108782         37  Regular customer
29418589    82650         42    Loyal customer
14723883   176047         40  Regular customer
27882268   112005         38  Regular customer
25846560    90731         32  Regular customer
9840264    115568         19  Regular customer
6211496    126656          8      New customer
12551193     3623         20  Regular customer
31179777   194971         54    Loyal customer
928814     137575         24  Regular customer
823200     152099         27  Regular customer
8140701    169583         55    Loyal customer
1299067     98389         36  Regular customer
20165180    38082         10      New customer
22403341    21384         34  Regular customer
21743766    38677         23  Regular customer
23802026    28826         50    Loyal customer
1379345    189834         27  Regular customer
29055797   168383         27  Regular customer
13199700   179737          7      New customer
15812331    26603         37  Regular customer
28146094    73893         14  Regular customer
28941543   116053         35  Regular customer
15350375   205966         32  Regular customer
26163197   134324         26  Regular customer
19846084   106749          3      New customer
16413632   201937         20  Regular customer
10014234    65389         34  Regular customer

Step 5
The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. 
Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category.
What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [37]:
print('The mean, min and max of prices grouped by the loyalty_flag column in ords_prods_merge:')
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in ords_prods_merge:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

1. It is intriguing to observe that new customers have a higher average purchase price compared to the other two customer categories, while even loyal customers have a lower average purchase price than the others.

2. The presence of an unusual maximum price of 99999 has significantly skewed the distribution of the mean value. Further analysis will be conducted to investigate and better understand this value.

In [38]:
print('Descriptive statistics of the prices column:')
ords_prods_merge['prices'].describe()

Descriptive statistics of the prices column:


count    3.243421e+07
mean     1.198114e+01
std      4.954311e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

The standard deviation confirms that there is a lot of variation in the data, especially with some very high values that are much higher than the average.

#Check for abnormal values in the 'prices' column


In [40]:
ords_prods_merge['prices'].sort_values(ascending = False)

29192983    99999.0
29192963    99999.0
29192971    99999.0
29192970    99999.0
29192969    99999.0
             ...   
23563370        1.0
23563371        1.0
23563372        1.0
23563373        1.0
7882813         1.0
Name: prices, Length: 32434212, dtype: float64

In [41]:
print('Check for abnormal values in the prices column:')

ords_prods_merge['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in the prices column:


29192575    99999.0
10030345    14900.0
18529046       25.0
25785942       24.9
28452491       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

Observation - 
The 99.999 and 14.900 prices appear unusual as they are significantly higher than the other prices in the dataset.


In [43]:
# Get the output of ords_prods_merge when the value in the 'prices' column is 99999 or 14900

max_price = ords_prods_merge.loc[ords_prods_merge['prices'].isin([99.999, 14.900])]

In [44]:
max_price.sample(10)

#below is the output of ords_prods_merge  when the value in the prices column is either 99.999 or 14.900

order_id  user_id eval_set  order_number  orders_day_of_week  \
10034240   1855072   181871    prior             3                   4   
10032965   1708474   122318    prior             8                   2   
10031395   1604943    54122    prior            12                   6   
10031732   3307117    68871    prior            36                   4   
10034351   3047652   186955    prior            28                   1   
10033733   2763053   159704    prior            19                   3   
29193028   1636564   147580    prior            31                   3   
29192892   1923734   108441    prior            15                   3   
10031891   2812144    74759    prior             3                   1   
10033279   1262568   138767    prior             3                   5   

          order_hour_of_day  days_since_prior_order  product_id  \
10034240                 19                    25.0       21553   
10032965                  8                     4.0       21553   
10031395                 12                     4.0       21553   
10031732                 17                     7.0       21553   
10034351                 14                     7.0       21553   
10033733                 17                    10.0       21553   
29193028                 23                    17.0       33664   
29192892                 19                     6.0       33664   
10031891                 10                     7.0       21553   
10033279                 17                    29.0       21553   

          add_to_cart_order  reordered                      product_name  \
10034240                 10          0  Lowfat 2% Milkfat Cottage Cheese   
10032965                  1          0  Lowfat 2% Milkfat Cottage Cheese   
10031395                 10          0  Lowfat 2% Milkfat Cottage Cheese   
10031732                  1          1  Lowfat 2% Milkfat Cottage Cheese   
10034351                  6          1  Lowfat 2% Milkfat Cottage Cheese   
10033733                 10          1  Lowfat 2% Milkfat Cottage Cheese   
29193028                 10          1             2 % Reduced Fat  Milk   
29192892                  7          1             2 % Reduced Fat  Milk   
10031891                  7          1  Lowfat 2% Milkfat Cottage Cheese   
10033279                 14          0  Lowfat 2% Milkfat Cottage Cheese   

          aisle_id  department_id   prices _merge     busiest_day  \
10034240       108             16  14900.0   both      Least busy   
10032965       108             16  14900.0   both  Regularly busy   
10031395       108             16  14900.0   both  Regularly busy   
10031732       108             16  14900.0   both      Least busy   
10034351       108             16  14900.0   both  Regularly busy   
10033733       108             16  14900.0   both  Regularly busy   
29193028        84             16  99999.0   both  Regularly busy   
29192892        84             16  99999.0   both  Regularly busy   
10031891       108             16  14900.0   both  Regularly busy   
10033279       108             16  14900.0   both  Regularly busy   

             busiest_days busiest_period_of_day  max_order      loyalty_flag  
10034240  Least busy days        Average orders         19  Regular customer  
10032965   Regularly busy        Average orders         28  Regular customer  
10031395   Regularly busy           Most orders         14  Regular customer  
10031732  Least busy days        Average orders         44    Loyal customer  
10034351     Busiest days           Most orders         44    Loyal customer  
10033733  Least busy days        Average orders         35  Regular customer  
29193028  Least busy days         Fewest orders         38  Regular customer  
29192892  Least busy days        Average orders         87    Loyal customer  
10031891     Busiest days           Most orders         22  Regular customer  
10033279   Regularly busy        Average orders         36  Regular cu

In [45]:
# Descriptive statistics on max_price

print('Descriptive statistics on max_price:')
max_price.describe()


Descriptive statistics on max_price:


order_id        user_id  order_number  orders_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000         5127.000000   
mean   1.708876e+06  106117.185684     15.302126            2.750536   
std    9.838956e+05   58594.982659     16.691849            2.010892   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.642320e+05   58136.000000      4.000000            1.000000   
50%    1.693799e+06  106021.000000      9.000000            3.000000   
75%    2.567244e+06  156389.000000     20.000000            5.000000   
max    3.421047e+06  206049.000000     99.000000            6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        5127.000000             4776.000000   5127.000000   
mean           13.473376               11.899079  23201.815682   
std             4.137326                9.068019   4153.742747   
min             0.000000                0.000000  21553.000000   
25%            10.000000                5.000000  21553.000000   
50%            13.000000                8.000000  21553.000000   
75%            16.000000               16.000000  21553.000000   
max            23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        5127.000000  5127.000000  5127.000000         5127.0   
mean            7.986347     0.613614   104.732592           16.0   
std             6.868889     0.486968     8.231346            0.0   
min             1.000000     0.000000    84.000000           16.0   
25%             3.000000     0.000000   108.000000           16.0   
50%             6.000000     1.000000   108.000000           16.0   
75%            11.000000     1.000000   108.000000           16.0   
max            72.000000     1.000000   108.000000           16.0   

             prices    max_order  
count   5127.000000  5127.000000  
mean   26485.547494    29.791691  
std    29186.636448    24.929207  
min    14900.000000     3.000000  
25%    14900.000000    12.000000  
50%    14900.000000    21.000000  
75%    14900.000000    41.000000  
max    99999.000000    99.000000

Observation: There are two products with extremely high prices (14900 and 99999) in the dataset. Both of these products belong to department_id 16 and have product_ids 21553 and 33664. These prices stand out as outliers compared to the rest of the data.

In [46]:
# Find the product_name with product_id 21553 and 33664

ords_prods_merge.loc[ords_prods_merge['product_id'].isin([21553, 33664])]

order_id  user_id eval_set  order_number  orders_day_of_week  \
10030345    912404       17    prior            12                   2   
10030346    603376       17    prior            22                   6   
10030347   3264360      135    prior             2                   2   
10030348    892534      135    prior             3                   0   
10030349    229704      342    prior             8                   1   
...            ...      ...      ...           ...                 ...   
29193268   2249946   204099    prior            29                   0   
29193269   2363282   204099    prior            31                   0   
29193270   3181945   204395    prior            13                   3   
29193271   2486215   205227    prior             7                   3   
29193272   2232716   205227    prior            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29193268                  8                     4.0       33664   
29193269                  9                     2.0       33664   
29193270                 15                     8.0       33664   
29193271                 20                     4.0       33664   
29193272                 11                     4.0       33664   

          add_to_cart_order  reordered                      product_name  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...                               ...   
29193268                  1          0             2 % Reduced Fat  Milk   
29193269                  1          1             2 % Reduced Fat  Milk   
29193270                 25          0             2 % Reduced Fat  Milk   
29193271                  8          0             2 % Reduced Fat  Milk   
29193272                 23          1             2 % Reduced Fat  Milk   

          aisle_id  department_id   prices _merge     busiest_day  \
10030345       108             16  14900.0   both  Regularly busy   
10030346       108             16  14900.0   both  Regularly busy   
10030347       108             16  14900.0   both  Regularly busy   
10030348       108             16  14900.0   both     Busiest day   
10030349       108             16  14900.0   both  Regularly busy   
...            ...            ...      ...    ...             ...   
29193268        84             16  99999.0   both     Busiest day   
29193269        84             16  99999.0   both     Busiest day   
29193270        84             16  99999.0   both  Regularly busy   
29193271        84             16  99999.0   both  Regularly busy   
29193272        84             16  99999.0   both     Busiest day   

             busiest_days busiest_period_of_day  max_order      loyalty_flag  
10030345   Regularly busy           Most orders         40  Regular customer  
10030346   Regularly busy           Most orders         40  Regular customer  
10030347   Regularly busy        Average orders          4      New customer  
10030348     Busiest days        Average orders          4      New customer  
10030349     Busiest days        Average orders         16  Regular customer  
...                   ...                   ...        ...               ...  
29193268     Busiest 

Observation:

The price of 14.900 is associated with the product "Lowfat 2% Milkfat Cottage Cheese" (product_id: 21553).
The price of 99.999 is associated with the product "2% Reduced Fat Milk" (product_id: 33664).

In [47]:
#Filter the dataset to include only records with department_id equal to 16:
#Create a filter to exclude records with very high prices (e.g., prices above 100):
#The ~ symbol is used to negate the condition, so it excludes the records that satisfy the condition. 
#In this case, it excludes records with prices above 100.

dairy = ords_prods_merge.loc[ords_prods_merge['department_id'].isin([16])]
dairy_without_weird = dairy.loc[~dairy['prices'].isin([99999, 14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
dairy_without_weird.sample(5)

Filter for records with the department_id value of 16, while excluding records without the unusually high prices:


order_id  user_id eval_set  order_number  orders_day_of_week  \
24918476   3087222   154592    prior             4                   0   
13319448    432149    57518    prior            18                   3   
27243310   1633187   186006    prior            81                   1   
21859483   1075890    71883    prior            15                   3   
25432981    927430   130481    prior             4                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
24918476                 13                     7.0       15404   
13319448                  9                     3.0       17630   
27243310                 10                     4.0       17316   
21859483                 15                    30.0       30949   
25432981                  8                     6.0       16414   

          add_to_cart_order  reordered  \
24918476                 17          1   
13319448                  5          1   
27243310                  8          1   
21859483                  6          0   
25432981                 12          0   

                                               product_name  aisle_id  \
24918476                            Fresh Sliced Mozzarella        21   
13319448   YoKids Squeeze! Organic Strawberry Flavor Yogurt       120   
27243310                       Vitamin D Organic Whole Milk        84   
21859483  Organic YoKids Smoothie Strawberry Banana Lowf...       120   
25432981                    Finely Shredded Parmesan Cheese        21   

          department_id  prices _merge     busiest_day     busiest_days  \
24918476             16    12.2   both     Busiest day     Busiest days   
13319448             16    14.7   both  Regularly busy  Least busy days   
27243310             16     2.9   both  Regularly busy     Busiest days   
21859483             16     3.5   both  Regularly busy  Least busy days   
25432981             16     3.9   both  Regularly busy     Busiest days   

         busiest_period_of_day  max_order      loyalty_flag  
24918476           Most orders         41    Loyal customer  
13319448           Most orders         85    Loyal customer  
27243310           Most orders         99    Loyal customer  
21859483           Most orders         16  Regular customer  
25432981        Average orders         21  Regular customer

In [48]:
# Check to see if the filter is working properly

print('Maximum price in the dairy_without_weird filter:')
dairy_without_weird['prices'].max()

Maximum price in the dairy_without_weird filter:


15.0

Observation: Upon excluding the unusual values, the maximum price in the dataset appears to be more reasonable and in line with the other prices. This suggests that the previously observed outliers had a significant impact on the maximum price value

In [52]:
# First: Observation on 2 % Reduced Fat Milk (with product_id 33664)
# Create a filter for products that contain 'Reduced Fat', '2' and 'Milk' in the name, in dairy_without_weird

reduced_fat_milk = dairy_without_weird[dairy_without_weird['product_name'].str.contains("Reduced Fat") & 
                                       dairy_without_weird['product_name'].str.contains("2") &
                                       dairy_without_weird['product_name'].str.contains("Milk")]

In [53]:
reduced_fat_milk.sample(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
16032087   1891189   109746    prior            17                   4   
11697015     71917    47955    prior             2                   4   
14559676   1859887    82872    prior            20                   2   
28639953   2427298   167032    prior             5                   6   
20186706   1935787    25533    prior             7                   2   
11702403   1599857   112196    prior             1                   1   
14560289   2106786    86796    prior             6                   6   
14551458     82720    38166    prior             6                   3   
14555993    365561    63428    prior            26                   3   
15487731   3193972   128257    prior             4                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
16032087                 16                    30.0        5785   
11697015                 14                    17.0       32478   
14559676                 18                    18.0       23909   
28639953                 15                    30.0       37795   
20186706                  9                    15.0       13166   
11702403                 16                     NaN       32478   
14560289                 17                    28.0       23909   
14551458                  1                    30.0       23909   
14555993                 10                    19.0       23909   
15487731                 18                    18.0       22610   

          add_to_cart_order  reordered                          product_name  \
16032087                  1          1           Organic Reduced Fat 2% Milk   
11697015                  1          1                   Reduced Fat 2% Milk   
14559676                  1          1                   2% Reduced Fat Milk   
28639953                  2          1        Dairy Pure 2% Reduced Fat Milk   
20186706                  1          1  Organic Milk Reduced Fat, 2% Milkfat   
11702403                  2          0                   Reduced Fat 2% Milk   
14560289                  7          0                   2% Reduced Fat Milk   
14551458                  2          1                   2% Reduced Fat Milk   
14555993                  5          1                   2% Reduced Fat Milk   
15487731                  1          1      Reduced Fat 2% Lactose-Free Milk   

          aisle_id  department_id  prices _merge     busiest_day  \
16032087        84             16     7.9   both      Least busy   
11697015        84             16    13.9   both      Least busy   
14559676        84             16     9.2   both  Regularly busy   
28639953        84             16    10.1   both  Regularly busy   
20186706        84             16    12.8   both  Regularly busy   
11702403        84             16    13.9   both  Regularly busy   
14560289        84             16     9.2   both  Regularly busy   
14551458        84             16     9.2   both  Regularly busy   
14555993        84             16     9.2   both  Regularly busy   
15487731        91             16     9.4   both      Least busy   

             busiest_days busiest_period_of_day  max_order      loyalty_flag  
16032087  Least busy days           Most orders         18  Regular customer  
11697015  Least busy days           Most orders          9      New customer  
14559676   Regularly busy        Average orders         23  Regular customer  
28639953   Regularly busy           Most orders          7      New customer  
20186706   Regularly busy           Most orders         61    Loyal customer  
11702403     Busiest days           Most orders         17  Regular customer  
14560289   Regularly busy        Average orders         10      New customer  
14551458  Least busy days         Fewest orders          8      New customer  
14555993  Least busy days           Most orders         32  Regular customer  
15487731  Least busy days        Aver

In [54]:
# statistics for the reduced_fat_milk

print('Quick statistics in reduced_fat_milk:')
reduced_fat_milk.describe()

Quick statistics in reduced_fat_milk:


order_id        user_id   order_number  orders_day_of_week  \
count  2.022380e+05  202238.000000  202238.000000       202238.000000   
mean   1.711414e+06  103040.384151      18.528684            2.748702   
std    9.876774e+05   59620.791055      18.288088            2.022770   
min    5.000000e+00      11.000000       1.000000            0.000000   
25%    8.586890e+05   51496.000000       5.000000            1.000000   
50%    1.706996e+06  102925.000000      12.000000            3.000000   
75%    2.566178e+06  154743.000000      26.000000            5.000000   
max    3.421036e+06  206209.000000      99.000000            6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      202238.000000           190556.000000  202238.000000   
mean           13.181835               10.538445   18768.761009   
std             4.193794                8.380235   13345.840890   
min             0.000000                0.000000    1940.000000   
25%            10.000000                5.000000    5785.000000   
50%            13.000000                7.000000   19820.000000   
75%            16.000000               14.000000   32478.000000   
max            23.000000               30.000000   49319.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      202238.000000  202238.000000  202238.000000       202238.0   
mean            5.423051       0.782692      83.716626           16.0   
std             5.397505       0.412415       6.511429            0.0   
min             1.000000       0.000000      21.000000           16.0   
25%             2.000000       1.000000      84.000000           16.0   
50%             4.000000       1.000000      84.000000           16.0   
75%             7.000000       1.000000      84.000000           16.0   
max           137.000000       1.000000      91.000000           16.0   

              prices      max_order  
count  202238.000000  202238.000000  
mean        9.086900      35.834794  
std         3.064869      26.069087  
min         1.200000       3.000000  
25%         7.900000      14.000000  
50%         9.200000      30.000000  
75%        11.400000      51.000000  
max        14.700000      99.000000

Observation: Mean, min, and max prices for 'Reduced Fat', '2', and 'Milk' products are now 9.08, 1.2, and 14.7, respectively. Given this, it is recommended to reduce the price of product '2% Reduced Fat Milk', from 99.999 to 9.99.

In [56]:
# Second: Observation on Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
# Create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird

cottage_cheese = dairy_without_weird[
    dairy_without_weird['product_name'].str.contains("Cottage Cheese") &
    dairy_without_weird['product_name'].str.contains("2") &
    dairy_without_weird['product_name'].str.contains("Milkfat")
]

In [57]:


print('Sample output from cottage_cheese:')
cottage_cheese.sample(10)



Sample output from cottage_cheese:


order_id  user_id eval_set  order_number  orders_day_of_week  \
16201480   2391226    11934    prior            20                   6   
29926367   1971635   168326    prior             1                   2   
29926023   1556232    43139    prior             1                   2   
16203285   2477154   167998    prior            19                   2   
16203376   2796752   174758    prior            16                   5   
30605105   2215798    99045    prior            30                   5   
16203577    740514   191659    prior            16                   3   
30604973   2910579     9825    prior            24                   5   
16203352   2349060   172435    prior            11                   1   
29926476   2907080   203992    prior             4                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
16201480                  8                    16.0       27790   
29926367                 13                     NaN        6852   
29926023                  6                     NaN        6852   
16203285                  9                    11.0       27790   
16203376                 16                     7.0       27790   
30605105                 17                     7.0       11203   
16203577                 12                     9.0       27790   
30604973                 17                    21.0       11203   
16203352                 13                    15.0       27790   
29926476                 12                    30.0        6852   

          add_to_cart_order  reordered  \
16201480                  2          1   
29926367                  5          0   
29926023                  3          0   
16203285                  2          1   
16203376                  3          0   
30605105                 21          1   
16203577                  5          1   
30604973                  6          1   
16203352                  1          1   
29926476                  8          1   

                                               product_name  aisle_id  \
16201480        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29926367         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   
29926023         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   
16203285        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16203376        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
30605105  Cottage Doubles Peach, 2% Milkfat Lowfat Cotta...       108   
16203577        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
30604973  Cottage Doubles Peach, 2% Milkfat Lowfat Cotta...       108   
16203352        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29926476         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   

          department_id  prices _merge     busiest_day     busiest_days  \
16201480             16     7.7   both  Regularly busy   Regularly busy   
29926367             16     8.6   both  Regularly busy   Regularly busy   
29926023             16     8.6   both  Regularly busy   Regularly busy   
16203285             16     7.7   both  Regularly busy   Regularly busy   
16203376             16     7.7   both  Regularly busy   Regularly busy   
30605105             16     2.4   both  Regularly busy   Regularly busy   
16203577             16     7.7   both  Regularly busy  Least busy days   
30604973             16     2.4   both  Regularly busy   Regularly busy   
16203352             16     7.7   both  Regularly busy     Busiest days   
29926476             16     8.6   both     Busiest day     Busiest days   

         busiest_period_of_day  max_order      loyalty_flag  
16201480        Average orders         20  Regular customer  
29926367           Most orders          4      New customer  
29926023         Fewest orders          3      New customer  
16203285           Most orders         58    Loyal customer  
16203376           Most orders         51    Loyal cus

In [58]:
# Quick statistics in cottage_cheese

print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


order_id        user_id  order_number  orders_day_of_week  \
count  3.487000e+03    3487.000000   3487.000000         3487.000000   
mean   1.684106e+06  103778.688271     15.484084            2.849728   
std    9.785618e+05   58583.418992     16.808776            2.031540   
min    2.301000e+03      52.000000      1.000000            0.000000   
25%    8.438150e+05   55839.000000      4.000000            1.000000   
50%    1.673171e+06  107636.000000     10.000000            3.000000   
75%    2.535750e+06  151541.500000     20.000000            5.000000   
max    3.419042e+06  206072.000000     98.000000            6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        3487.000000             3218.000000   3487.000000   
mean           13.257241               12.898073  22348.791225   
std             4.297401                9.231128   8826.332190   
min             0.000000                0.000000    195.000000   
25%            10.000000                6.000000  11203.000000   
50%            13.000000               10.000000  27790.000000   
75%            16.000000               19.000000  27790.000000   
max            23.000000               30.000000  31030.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        3487.000000  3487.000000  3487.000000         3487.0   
mean            5.840551     0.697448   107.650703           16.0   
std             5.955836     0.459429     5.502322            0.0   
min             1.000000     0.000000    21.000000           16.0   
25%             2.000000     0.000000   108.000000           16.0   
50%             4.000000     1.000000   108.000000           16.0   
75%             8.000000     1.000000   108.000000           16.0   
max            57.000000     1.000000   108.000000           16.0   

            prices    max_order  
count  3487.000000  3487.000000  
mean      7.228219    30.531116  
std       1.757900    25.413902  
min       2.400000     3.000000  
25%       7.700000    11.000000  
50%       7.700000    22.000000  
75%       7.700000    43.000000  
max      11.800000    99.000000

Observation: The products that contain 'Cottage Cheese', '2', and 'Milkfat' in the name have an average price of 7.2, with a minimum price of 2.4 and a maximum price of 11.8. Since this result makes more sense, the price of product ID 21553, named 'Lowfat 2% Milkfat Cottage Cheese', will be replaced with the mean value of 7.23.

#Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese

In [59]:
ords_prods_merge['prices'] = ords_prods_merge['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})


In [61]:
ords_prods_merge['prices'].describe()

count    3.243421e+07
mean     7.795674e+00
std      4.242831e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

step - 5

In [62]:
# Basic statistics of ords_prods_merge by 'loyalty_flag' with the updated 'prices' column

print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:')
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.778386  1.0  25.0
New customer      7.805644  1.0  25.0
Regular customer  7.802952  1.0  25.0

Observation: When grouping the prices by loyalty_flag in the ords_prods_merge DataFrame, the mean value has decreased, and the maximum value is now more reasonable at 25 instead of 99999. This indicates that the updated prices have influenced the overall statistics based on the loyalty_flag groups.

STEP - 6 The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [ ]:
# Creating “average_price” column using transform()

In [63]:
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [64]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... aisle_id  department_id  prices  _merge     busiest_day  \
0          0  ...       77              7     9.0    both  Regularly busy   
1          1  ...       77              7     9.0    both  Regularly busy   
2          1  ...       77              7     9.0    both  Regularly busy   
3          1  ...       77              7     9.0    both      Least busy   
4          1  ...       77              7     9.0    both      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag average_price  
0   Regularly busy        Average orders        10  New customer      6.367797  
1  Least busy days        Average orders        10  New customer      6.367797  
2  Least busy days           Most orders        10  New customer      6.367797  
3  Least busy days        Average orders        10  New customer      6.367797  
4  Least busy days           Most orders        10  New customer      6.367797  

[5 rows x 21 columns]

In [65]:
# Rounding averages 

ords_prods_merge['average_price'] = ords_prods_merge['average_price'].round(2)

In [66]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [67]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [68]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... department_id  prices  _merge     busiest_day  \
0          0  ...             7     9.0    both  Regularly busy   
1          1  ...             7     9.0    both  Regularly busy   
2          1  ...             7     9.0    both  Regularly busy   
3          1  ...             7     9.0    both      Least busy   
4          1  ...             7     9.0    both      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1  Least busy days        Average orders        10  New customer   
2  Least busy days           Most orders        10  New customer   
3  Least busy days        Average orders        10  New customer   
4  Least busy days           Most orders        10  New customer   

   average_price spending_flag  
0           6.37   Low spender  
1           6.37   Low spender  
2           6.37   Low spender  
3           6.37   Low spender  
4           6.37   Low spender  

[5 rows x 22 columns]

In [69]:


# Check the frequency

ords_prods_merge['spending_flag'].value_counts(dropna = False)



Low spender     32312862
High spender      121350
Name: spending_flag, dtype: int64


Step 7 In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


In [14]:
# Creating “median_prior_orders” column using transform()
ords_prods_merge['median_prior_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [15]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

      busiest_day     busiest_days busiest_period_of_day  median_prior_orders  
0  Regularly busy   Regularly busy        Average orders                 20.5  
1  Regularly busy  Least busy days        Average orders                 20.5  
2  Regularly busy  Least busy days           Most orders                 20.5  
3      Least busy  Least busy days        Average orders                 20.5  
4      Least busy  Least busy days           Most orders                 20.5

In [16]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [17]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[(ords_prods_merge['median_prior_orders'] > 10) & (ords_prods_merge['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [18]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [19]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

      busiest_day     busiest_days busiest_period_of_day  median_prior_orders  \
0  Regularly busy   Regularly busy        Average orders                 20.5   
1  Regularly busy  Least busy days        Average orders                 20.5   
2  Regularly busy  Least busy days           Most orders                 20.5   
3      Least busy  Least busy days        Average orders                 20.5   
4      Least busy  Least busy days           Most orders                 20.5   

    order_frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer

#Check the frequency

In [22]:
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21577409
Regular customer          7217134
Non-frequent customer     3639669
Name: order_frequency_flag, dtype: int64

In [23]:
# Check the null values

print('Records in df_orders_products_merged with null values in column order_frequency_flag:')
ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'median_prior_orders', 'order_frequency_flag']]

Records in df_orders_products_merged with null values in column order_frequency_flag:


Empty DataFrame
Columns: [user_id, days_since_prior_order, median_prior_orders, order_frequency_flag]
Index: []

In [21]:
# Check user_id 159838

ords_prods_merge[ords_prods_merge['user_id'] == 159838]

order_id  user_id eval_set  order_number  orders_day_of_week  \
13652228    895835   159838    prior             1                   0   
16346534    895835   159838    prior             1                   0   
16346535    707970   159838    prior             3                   1   
16346536   3275273   159838    prior             4                   2   
16346537   2922254   159838    prior             6                   4   
17276282    895835   159838    prior             1                   0   
17647059    895835   159838    prior             1                   0   
24162885    895835   159838    prior             1                   0   
25907061    895835   159838    prior             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13652228                 17                     NaN       10749   
16346534                 17                     NaN        1511   
16346535                 12                    30.0        1511   
16346536                 15                    15.0        1511   
16346537                 13                    19.0        1511   
17276282                 17                     NaN       33401   
17647059                 17                     NaN       23695   
24162885                 17                     NaN       21334   
25907061                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13652228                  3          0   
16346534                  4          0   
16346535                  2          1   
16346536                  2          1   
16346537                  2          1   
17276282                  6          0   
17647059                  2          0   
24162885                  5          0   
25907061                  1          0   

                                               product_name  aisle_id  \
13652228                            Organic Red Bell Pepper        83   
16346534                                                NaN        84   
16346535                                                NaN        84   
16346536                                                NaN        84   
16346537                                                NaN        84   
17276282                               Goat Cheese Crumbles        21   
17647059                           California Veggie Burger        42   
24162885                              Organic Peeled Garlic       123   
25907061  4X Ultra Concentrated Natural Laundry Detergen...        75   

          department_id  prices _merge     busiest_day     busiest_days  \
13652228              4     5.7   both     Busiest day     Busiest days   
16346534             16    14.3   both     Busiest day     Busiest days   
16346535             16    14.3   both  Regularly busy     Busiest days   
16346536             16    14.3   both  Regularly busy   Regularly busy   
16346537             16    14.3   both      Least busy  Least busy days   
17276282             16    14.8   both     Busiest day     Busiest days   
17647059              1     4.7   both     Busiest day     Busiest days   
24162885              4    10.2   both     Busiest day     Busiest days   
25907061             17     1.7   both     Busiest day     Busiest days   

         busiest_period_of_day  median_prior_orders order_frequency_flag  
13652228        Average orders                 19.0     Regular customer  
16346534        Average orders                 19.0     Regular customer  
16346535           Most orders                 19.0     Regular customer  
16346536           Most orders                 19.0     Regular customer  
16346537           Most orders                 19.0     Regular customer  
17276282        Average orders                 19.0     Regular customer  
17647059        Average orders                 19.0     Regular customer  
24162885        Average orders                 19.0     Regular customer  
25907061        Aver

Observation: The user_id 159838 appears to be a new customer, and the NaN values in the column 'days_since_prior_order'are left unchanged.

In [24]:
ords_prods_merge.shape

(32434212, 20)

step - 8 Ensure your notebook is clean and structured and that your code is well commented.


#Done

step - 9 Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [6]:
ords_prods_merge.to_pickle(os.path.join(path, 'orders_products_merged_updated_2.pkl' ))